### PR TITLE
Add an option to use the X-B3-Flags header to override any sampling d…

### LINKED ISSUE #496: Add an option to use the X-B3-Flags header to override any sampling decision
without this change it's pretty much impossible to enforce sampling for certain traces
with this change setting the `X-B3-Flags` to `1` for HTTP messages / `spanFlags` to `1` for messaging will override any sampling decisions

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
…ecision

without this change it's pretty much impossible to enforce sampling for certain traces
with this change setting the X-B3-Flags to 1 for HTTP messages / spanFlags to 1 for messaging will override any sampling decisions

fixes #496

cc @adriancole

### CODE REVIEW

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractorTests.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+import static org.springframework.cloud.sleuth.assertions.SleuthAssertions.then;
+
+/**
+ * @author Marcin Grzejszczak
+ */
+public class HeaderBasedMessagingExtractorTests {
+
+	@Test
+	public void overridesTheSampleFlagWithSpanFlagForSampledScenario() {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractorTests.java]: maybe add a test for when only the flags header is present? Here's the code from finagle..

```
    } else if (request.headers.contains(Header.Flags)) {
      // even if there are no id headers we want to get the debug flag
      // this is to allow developers to just set the debug flag to ensure their
      // trace is collected
      Some(Trace.nextId.copy(flags = getFlags(request)))
```

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java [hunk 2]
```diff
@@ -382,6 +382,38 @@ public void returns200IfSpanParentIsMalformedAndCreatesANewSpan() throws Excepti
...
+		TraceFilter filter = new TraceFilter(this.tracer, this.traceKeys, this.spanReporter,
+				this.spanExtractor, this.httpTraceKeysInjector);
+
+		filter.doFilter(this.request, this.response, this.filterChain);
+
+		then(new ListOfSpans(this.spanReporter.getSpans())).allSpansAreExportable();
+		then(TestSpanContextHolder.getCurrentSpan()).isNull();
+		then(ExceptionUtils.getLastException()).isNull();
+	}
+
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: there's a subtle side-effect on debug not in this change. that is that it sets Span.debug=true. This makes sure the collector doesn't throw the span away either.. (note this only shows up when the server has collector sampling as a rate less than 1.0)

Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: ah so you're saying that we need to update the collectors to check for the presence of that flag?

**Comment by adriancole** [comment]
>
> ah so you're saying that we need to update the collectors to check for the
> presence of that flag?
>
yep or add a TODO and wait for someone to ask :P

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java [hunk 3]
```diff
@@ -414,6 +414,42 @@ public void doesNotOverrideTheSampledFlagWhenXB3FlagIsSetToOtherValueThan1() thr
...
+				.allSpansAreExportable().hasSize(2).hasASpanWithSpanId(Span.hexToId("10"));
+		then(TestSpanContextHolder.getCurrentSpan()).isNull();
+		then(ExceptionUtils.getLastException()).isNull();
+	}
+
+	@Test
+	public void samplesWhenDebugFlagIsSetTo1AndOnlyTraceIdIsSet() throws Exception {
+		this.request = builder()
+				.header(Span.SPAN_FLAGS, 1)
+				.header(Span.TRACE_ID_NAME, 10L)
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: heh mild lie :) trace id name is also set!

Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: lol - indeed ;)

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractor.java [hunk 4]
```diff
@@ -17,14 +18,23 @@
 	@Override
 	public Span joinTrace(SpanTextMap textMap) {
 		Map<String, String> carrier = TextMapUtil.asMap(textMap);
-		if (!hasHeader(carrier, TraceMessageHeaders.SPAN_ID_NAME)
+		if (Span.SPAN_SAMPLED.equals(carrier.get(TraceMessageHeaders.SPAN_FLAGS_NAME))) {
+			generateIdIfMissing(carrier, TraceMessageHeaders.SPAN_ID_NAME);
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractor.java]: this is fine, though you could make it simpler by using the trace == span id when absent

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/ZipkinHttpSpanExtractor.java [hunk 5]
```diff
@@ -33,7 +34,10 @@ public ZipkinHttpSpanExtractor(Pattern skipPattern) {
 	@Override
 	public Span joinTrace(SpanTextMap textMap) {
 		Map<String, String> carrier = TextMapUtil.asMap(textMap);
-		if (carrier.get(Span.TRACE_ID_NAME) == null) {
+		boolean debug = Span.SPAN_SAMPLED.equals(carrier.get(Span.SPAN_FLAGS));
+		if (debug) {
+			generateIdIfMissing(carrier, Span.TRACE_ID_NAME);
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/ZipkinHttpSpanExtractor.java]: does this need to also set span id? if not probably add a comment why

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractor.java (hunk 4 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingExtractor.java]: True!

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/ZipkinHttpSpanExtractor.java (hunk 5 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/ZipkinHttpSpanExtractor.java]: TraceId is enough cause later we're generating a span if it's missing. I'll add the comment

**Comment by marcingrzejszczak** [comment]
done via https://github.com/spring-cloud/spring-cloud-sleuth/commit/c347f724cdb9cba96f89dbc0c828f99d614ebdfc

**Pull request closed without merging**